File: sdks/dotnet/sulfone-helium/Domain/Processor/Service.cs
using sulfone_helium.Domain.Core;
using sulfone_helium.Domain.Core.FileSystem;

namespace sulfone_helium.Domain.Processor;

public class ProcessorService
{
    public required ICyanProcessor Processor { private get; init; }

    public async Task<ProcessorOutput> Process(ProcessorInput input)
    {
        var (read, write, globs, config) = input;
        var cyanGlobs = globs as CyanGlob[] ?? globs.ToArray();
        var helper = new CyanFileHelper(read, write, cyanGlobs);
        return await Processor.Process(new CyanProcessorInput
        {
            ReadDir = read,
            WriteDir = write,
            Globs = cyanGlobs,
            Config = config,
        }, helper);
    }
}

File: sdks/node/src/domain/processor/service.ts
import type { ICyanProcessor } from "../core/cyan_script.js";
import type { ProcessorInput } from "./input.js";
import type { ProcessorOutput } from "./output.js";
import { CyanFileHelper } from "../core/fs/cyan_fs_helper.js";

class ProcessorService {
  private _processor: ICyanProcessor;

  constructor(processor: ICyanProcessor) {
    this._processor = processor;
  }

  async process(input: ProcessorInput): Promise<ProcessorOutput> {
    const { readDirectory, writeDirectory, globs, config } = input;
    const helper = new CyanFileHelper(readDirectory, writeDirectory, globs);
    return await this._processor.process(
      {
        readDir: readDirectory,
        writeDir: writeDirectory,
        globs,
        config: config,
      },
      helper,
    );
  }
}

export { ProcessorService };
